### ISSUE TITLE
Annotations -- should not affect ties

### ISSUE THREAD

**Issue by mscuthbert** [vexflow/vexflow#182]
(on 4.3.2)

When computing the starting point of a tie there is an issue if an annotation like lyrics exist that are pretty long.  The tie gets smaller and smaller as in the last notes here.

<img width="315" alt="Screenshot 2024-03-07 at 16 53 42" src="https://github.com/vexflow/vexflow/assets/3521479/a941fa13-ddb2-40b3-a8b2-352a7961237d">

Eventually, if the lyric is long enough, the tie vanishes completely:

<img width="557" alt="Screenshot 2024-03-07 at 16 56 50" src="https://github.com/vexflow/vexflow/assets/3521479/d5ece4b8-65b8-442c-8641-fc0b9d903790">

**Comment by mscuthbert** [comment]
Figured out a shim that works on this.  I'll happy contribute it into Vexflow 5, but I'd like comment on whether there's a danger to this approach which involves creating a new, disconnected modifierContext without lyrics, measuring it, and subtracting the difference.

```typescript
function getLyricWidthDifference(sn: StaveNote): {left: number, right: number} {
    const myModifierContext = sn.getModifierContext();
    myModifierContext.preFormat();  // just in case it hasn't been done yet.
    const myLeft = myModifierContext.getState().left_shift;
    const myRight = myModifierContext.getState().right_shift;

    const modifierContext2 = new ModifierContext();
    for (const cat of Object.keys((myModifierContext as any).members)) {
        if (cat === Annotation.CATEGORY) {
            continue;
        }
        for (const member of myModifierContext.getMembers(cat)) {
            modifierContext2.addMember(member);
        }
    }
    modifierContext2.preFormat();
    const noLyricLeft = modifierContext2.getState().left_shift;
    const noLyricRight = modifierContext2.getState().right_shift;
    return {
        left: myLeft - noLyricLeft,
        right: myRight - noLyricRight,
    };
}

const original_StaveNote_getTieRightX = StaveNote.prototype.getTieRightX;
StaveNote.prototype.getTieRightX = function getTieRightX(): number {
    const originalRight = original_StaveNote_getTieRightX.bind(this)();
    return originalRight - getLyricWidthDifference(this).right;
};
```

**Comment by AaronDavidNewman** [comment]
Wouldn't you have the same problem with modifiers that come after the note head, like ornaments?  It seems like you would.

It would be nice to make this more generic somehow.  It seems like there's other cases where you want to consider the X of only modifiers that are around the same line as the note head.  Maybe tie is a special case because the Y of both sides is always the same (although there can be different # of pitches on each side.

Maybe StaveTie always ignores/subtracts the annotation width if the vertical alignment is above or below the stave.

**Comment by mscuthbert** [comment]
> Wouldn't you have the same problem with modifiers that come after the note head, like ornaments? It seems like you would.

It depends on whether the ornaments are at a Y position that could intersect with the tie.  Trills would be no problem.  Gruppetti (Turns) might be a conflict.  The W3C MNX group (which I'm on as current editor of the MusicXML standard) has come to the conclusion that for rendering purposes we should not be grouping "modifiers" by their musical (=performance) characteristics but according to their impact on positioning (that is: all stemmodifiers grouped together, all notehead modifiers together, etc.)  So a trill that applied to the whole StaveNote would be displayed differently (usually above the note) and stored differently than a trill that applied to just one notehead/key (usually displayed left or right of the note). 


> It would be nice to make this more generic somehow. It seems like there's other cases where you want to consider the X of only modifiers that are around the same line as the note head. Maybe tie is a special case because the Y of both sides is always the same (although there can be different # of pitches on each side.

My larger idea is that preFormat on StaveNote (or StaveX...) should be broken up into two methods.  leave preFormat's input/output and setting state alone, but have it call a new method, which takes in a state and returns the modified state and also takes in an ordered list of Classes to call which affect the state.  In other words, we should be able to compute state without changing state.  If we took in an ordered list of classes to the "compute-but-don't-change" method, then we'd be able to figure out the bounding box for a note with an subset of modifiers, without changing anything in the underlying representation.

> Maybe StaveTie always ignores/subtracts the annotation width if the vertical alignment is above or below the stave.

I think that that's a good intermediate solution.  The more robust solution is to make Annotations (=text near notehead) a separate class from Lyric (=text below staff that doesn't interfere with ties).

I do think that even without doing that, this PR is an improvement.  If there's a non-lyric annotation near the note and someone is using ties weirdly, the worst problem is that the tie overlaps with that annotation, and it's hard to read.  Currently, if someone uses an annotation as a lyric (as is recommended) then the tie will shrink so much as to be non-existent to the user.  I think that scenario I is better than scenario II.

**Comment by AaronDavidNewman** [comment]
No, I think this is good.  I also use Annotations for lyrics in smoosic and I see the same issue with ties.

As for creating a 'dummy' modifiercontext, it could get out of hand if we started doing it with lots of modifiers.  But I think you should do it anyway :)  It may be the best way to handle the case where a modifier is formatted differently in the presence of other modifiers.  If we find ourselves doing it over and over, maybe we can think of a more wholistic solution.

**Comment by mscuthbert** [comment]
My thought was to factor out from preFormat the computation of the "state" values from the setting of "state" and to make the list of categories to compute be a modifiable array. The it becomes easy for an implementor to add new Modifier classes or remove certain ones. So I'll go that direction in the PR.
